ir/src/error.rs

```diff
@@ -74,4 +80,150 @@ impl SemanticError {
             matrix_col_len
         ))
     }
+
+    pub(super) fn named_trace_column_access_out_of_bounds(
+        access: &NamedTraceAccess,
+        size: usize,
+    ) -> Self {
+        SemanticError::IndexOutOfRange(format!(
+            "Out-of-range index '{}' while accessing named trace column group '{}' of length {}",
+            access.idx(),
+            access.name(),
+            size
+        ))
+    }
+
+    pub(super) fn indexed_trace_access_out_of_bounds(
+        access: &IndexedTraceAccess,
+        size: usize,
+    ) -> Self {
+        SemanticError::IndexOutOfRange(format!(
+            "Segment index '{}' is greater than the number of segments in the trace ({}).",
+            access.trace_segment(),
+            size
+        ))
+    }
+
+    pub(super) fn indexed_trace_column_access_out_of_bounds(
+        access: &IndexedTraceAccess,
+        segment_width: u16,
+    ) -> Self {
+        SemanticError::IndexOutOfRange(format!(
+            "Out-of-range index '{}' in trace segment '{}' of length {}",
+            access.col_idx(),
+            access.trace_segment(),
+            segment_width
+        ))
+    }
+
+    pub(super) fn random_value_access_out_of_bounds(index: usize, size: u16) -> Self {
+        SemanticError::IndexOutOfRange(format!(
+            "Random value index {index} is greater than or equal to the total number of random values ({size})."
+        ))
+    }
+
+    // --- DECLARATION ERRORS ---------------------------------------------------------------------
+
+    fn missing_declaration(missing_section: &str) -> Self {
+        SemanticError::MissingDeclaration(format!("{missing_section} section is missing"))
+    }
+
+    pub(super) fn missing_trace_columns_declaration() -> Self {
+        Self::missing_declaration("trace_declaration")
+    }
+
+    pub(super) fn missing_public_inputs_declaration() -> Self {
+        Self::missing_declaration("public_inputs")
+    }
+
+    pub(super) fn missing_boundary_constraints_declaration() -> Self {
+        Self::missing_declaration("boundary_constraints")
+    }
+
+    pub(super) fn missing_integrity_constraints_declaration() -> Self {
+        Self::missing_declaration("integrity_constraints")
+    }
+
+    pub(super) fn has_random_values_but_missing_aux_trace_columns_declaration() -> Self {
+        SemanticError::MissingDeclaration(
+            "random_values section requires aux_trace_columns section, which is missing"
+                .to_string(),
+        )
+    }
+
+    pub(super) fn illegal_identifer_type(ident_name: &str, ident_type: &IdentifierType) -> Self {
+        SemanticError::InvalidUsage(format!(
+            "Identifier {ident_name} was declared as a {ident_type} which is not a supported type."
+        ))
+    }
```

Comment:
should this be moved down under the "ILLEGAL IDENTIFIER ERRORS" header?

Reply:
Probably - I don't know the relationship between identifiers and their types, so I figured the type was part of the declaration of the identifier.

I'll move it.